pennylane/beta/plugins/default_tensor.py

```diff
@@ -572,73 +636,82 @@ def ev(self, obs_nodes, wires):
+                    input_idx = len(obs_wires) + idx
+                    self._add_edge(obs_node, input_idx, ket, w)  # A|psi>
+                    self._add_edge(bra, w, obs_node, output_idx)  # <psi|A
+            for w in set(all_wires) - set(meas_wires):
+                self._add_edge(bra, w, ket, w)  # |psi[w]|**2
+
+            # At this stage, all nodes are connected, and the contraction yields a
+            # scalar value.
+            contracted_ket = ket
+            for obs_node in obs_nodes:
+                contracted_ket = tn.contract_between(obs_node, contracted_ket)
+            expval = tn.contract_between(bra, contracted_ket).tensor
+
+        elif self._rep == "mps":
+            raise NotImplementedError
+            # if len(wires) == 1:
+            #    expval = self.mps.measure_local_operator(obs_nodes, wires[0])
+            # else:
+            #    raise NotImplementedError
+
+        if self._abs(self._imag(expval)) > TOL:
             warnings.warn(
                 "Nonvanishing imaginary part {} in expectation value.".format(expval.imag),
                 RuntimeWarning,
             )
         return self._real(expval)
 
+    def _contract_to_ket(self, method="auto"):
+        """Contract the nodes which represent the state preparation and evolution to a final tensor.
+        The contracted tensor is stored in the `_nodes` dictionary under the key `"contracted_state"`
+
+        Args:
+            method (str): the contraction method to be employed (possible choices are listed
+             in the dictionary `contract_fns`)
+        """
+        if "contracted_state" not in self._nodes:
+            contract = contract_fns[method]
+            ket = contract(self._nodes["state"], output_edge_order=self._terminal_edges)
+            ket.set_name("Ket")
+            self._nodes["contracted_state"] = ket
+
     @property
-    def _state(self):
+    def _state(self, contraction_method="auto"):
         """The numerical value of the current state vector.
 
         This attribute cannot be manually overwritten.
 
         Returns:
             (array, tf.Tensor, torch.Tensor): the numerical tensor
         """
-
-        return self._state_node.tensor
-
-    def reset(self):
-        """Reset the device"""
-        self._nodes = []
-        self._edges = []
-
-        state = self._create_basis_state([0] * self.num_wires, range(self.num_wires))
-        state = self._array(state, dtype=self.C_DTYPE)
-
-        # TODO: since this state is separable, can be more intelligent about not making a dense matrix
-        self._state_node = self._add_node(
-            state, wires=tuple(w for w in range(self.num_wires)), name="AllZeroState"
-        )
-        self._free_edges = self._state_node.edges[
-            :
-        ]  # we need this list to be distinct from self._state_node.edges
+        self._contract_to_ket(contraction_method)
+        ket = self._nodes["contracted_state"]
+        return ket.tensor
```

Comment:
Oh, I think I see (correct me if wrong):

* First time `_state` is called, the network is contracted, and the state stored in the `_nodes` dictionary.

* If `_state` is called again, the contracted state is retrieved from the dictionary rather than contracting again (since the network is _already_ contracted).

If so, perhaps it makes more sense to have that logic here, rather than in a else-less if statement within `_contract_to_ket`?

In fact, moving the logic here it becomes a one-liner (I think?) e.g.

```python
def _state(self, contraction_method="auto"):
    return self._nodes.setdefault("contracted_state", self._contract_to_ket(contraction_method))
```

Reply:
Just curious your thoughts on the suggestion here (can see it is marked resolved but not sure if any changes were made or it was left as is). It looks like the only place `self._contract_to_ket` is called, apart from here, is in `self.ev`.

By changing the logic to move the 'state caching' here rather than in `self._contract_to_ket`, this could also remove the state retrieval logic from `self.ev` as well, simply becoming

```python
if self._rep == "exact":
    ket = self._state(contraction_method)
```

Reply:
Rather than saving a couple lines of code, I didn't see enough benefit to switching this logic at the moment. (and when I tried to switch it, it introduced a whole bunch of failing tests)